sound/soc/sof/intel/hda.c

```diff
@@ -977,12 +979,13 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	/* create codec instances */
 	hda_codec_probe_bus(sdev);
 
-	if (!HDA_IDISP_CODEC(bus->codec_mask))
+err:
+	if (!HDA_IDISP_CODEC(bus->codec_mask) || i915_off)
 		hda_codec_i915_display_power(sdev, false);
```

Comment:
On other note: how this supposed to working in the original case??
```c
void hda_codec_i915_display_power(struct snd_sof_dev *sdev, bool enable)
{
	struct hdac_bus *bus = sof_to_bus(sdev);

	if (IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC_DEBUG_SUPPORT) &&
	    sof_debug_check_flag(SOF_DBG_FORCE_NOCODEC))
		return;

	if (HDA_IDISP_CODEC(bus->codec_mask)) {
		dev_dbg(bus->dev, "Turning i915 HDAC power %d\n", enable);
		snd_hdac_display_power(bus, HDA_CODEC_IDX_CONTROLLER, enable);
	}
}
EXPORT_SYMBOL_NS_GPL(hda_codec_i915_display_power, SND_SOC_SOF_HDA_AUDIO_CODEC_I915);
```
If we only call `hda_codec_i915_display_power()` in case `!HDA_IDISP_CODEC(bus->codec_mask)` then the whole thing is a NOP?

The power initially turned on via hda_init()->hda_codec_i915_init()->snd_hdac_display_power() directly and if the hda_iit_caps() fails it is going to be turned off via the error handling: hda_dsp_probe()->hda_codec_i915_exit()

Reply:
actually there are two problems with the code. First this:

````c
	/* scan SoundWire capabilities exposed by DSDT */
	ret = hda_sdw_acpi_scan(sdev);
	if (ret < 0) {
		dev_dbg(sdev->dev, "skipping SoundWire, not detected with ACPI scan\n");
		goto skip_soundwire;
	}
````

if ret < 0 then we would stop the probe. That's not the intended design and the idea with 'skip_soundwire'.

On error cases we do turn off the i915 as explained by @ujfalusi. The i915 power management is a bit misleading in the power is turned on/off in multiple places. It's unconditionally turned on/off in hda_codev_i915_init/exit and conditionally everyone else. that's fine, but @ujfalusi has a point that this is a no-op:

````
if (!HDA_IDISP_CODEC(bus->codec_mask))
		hda_codec_i915_display_power(sdev, false);
````

this code was added by @kv2019i in 0c75419a94a20b2bf791825a665d8b64cf7c560d, but our very own @kv2019i added an additional filter in 816938b272b0ac0203e25ce50483bd284ea4a2db which now makes the initial code a no-op.

One of the two changes is correct, which one @kv?

Reply:
I'll drop this patch and added https://github.com/thesofproject/linux/issues/3993 to track the i915 thread.